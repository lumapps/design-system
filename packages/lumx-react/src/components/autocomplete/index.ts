export * from './Autocomplete';
export * from './AutocompleteMultiple';
