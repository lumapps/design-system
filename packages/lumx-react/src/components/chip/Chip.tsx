import { Color, ColorPalette, Size, Theme } from '@lumx/react';
import { useStopPropagation } from '@lumx/react/hooks/useStopPropagation';

import { Comp, GenericProps, getRootClassName, handleBasicClasses, onEnterPressed } from '@lumx/react/utils';

import classNames from 'classnames';

import isFunction from 'lodash/isFunction';
import React, { forwardRef, MouseEventHandler, ReactNode } from 'react';

/**
 * Chip sizes.
 */
type ChipSize = Extract<Size, 's' | 'm'>;

/**
 * Defines the props of the component.
 */
export interface ChipProps extends GenericProps {
    /** A component to be rendered after the content. */
    after?: ReactNode;
    /** A component to be rendered before the content. */
    before?: ReactNode;
    /** Color variant. */
    color?: Color;
    /** Whether the component is clickable or not. */
    isClickable?: boolean;
    /** Whether the component is disabled or not. */
    isDisabled?: boolean;
    /** Whether the chip is currently in a highlighted state or not. */
    isHighlighted?: boolean;
    /** Whether the component is selected or not. */
    isSelected?: boolean;
    /** Size variant. */
    size?: ChipSize;
    /** Theme adapting the component to light or dark background. */
    theme?: Theme;
    /** On "after" element clicked callback. */
    onAfterClick?: MouseEventHandler;
    /** On "before" element clicked callback. */
    onBeforeClick?: MouseEventHandler;
}

/**
 * Component display name.
 */
const COMPONENT_NAME = 'Chip';

/**
 * Component default class name and class prefix.
 */
const CLASSNAME = getRootClassName(COMPONENT_NAME);

/**
 * Component default props.
 */
const DEFAULT_PROPS: Partial<ChipProps> = {
    size: Size.m,
    theme: Theme.light,
};

/**
 * Chip component.
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */
export const Chip: Comp<ChipProps, HTMLAnchorElement> = forwardRef((props, ref) => {
    const {
        after,
        before,
        children,
        className,
        color,
        disabled,
        isClickable,
        isDisabled = disabled,
        isHighlighted,
        isSelected,
        onAfterClick,
        onBeforeClick,
        onClick,
        size,
        theme,
        ...forwardedProps
    } = props;
    const hasAfterClick = isFunction(onAfterClick);
    const hasBeforeClick = isFunction(onBeforeClick);
    const hasOnClick = isFunction(onClick);

    // Adapt color to the theme.
    const chipColor = color || (theme === Theme.light ? ColorPalette.dark : ColorPalette.light);

    const handleOnBeforeClick = useStopPropagation(onBeforeClick);
    const handleOnAfterClick = useStopPropagation(onAfterClick);

    return (
        // eslint-disable-next-line jsx-a11y/no-static-element-interactions
        <a
            {...forwardedProps}
            ref={ref}
            className={classNames(
                className,
                handleBasicClasses({
                    clickable: Boolean(hasOnClick) || isClickable,
                    color: chipColor,
                    isDisabled,
                    hasAfter: Boolean(after),
                    hasBefore: Boolean(before),
                    highlighted: Boolean(isHighlighted),
                    prefix: CLASSNAME,
                    selected: Boolean(isSelected),
                    size,
                    unselected: Boolean(!isSelected),
                }),
            )}
            role={hasOnClick ? 'button' : undefined}
            tabIndex={isDisabled || !hasOnClick ? -1 : 0}
            aria-disabled={(hasOnClick && isDisabled) || undefined}
            onClick={hasOnClick ? onClick : undefined}
            onKeyDown={hasOnClick ? onEnterPressed(onClick) : undefined}
        >
            {before && (
                // eslint-disable-next-line jsx-a11y/click-events-have-key-events,jsx-a11y/no-static-element-interactions
                <div
                    className={classNames(`${CLASSNAME}__before`, {
                        [`${CLASSNAME}__before--is-clickable`]: hasBeforeClick,
                    })}
                    onClick={handleOnBeforeClick}
                >
                    {before}
                </div>
            )}

            <div className={`${CLASSNAME}__label`}>{children}</div>

            {after && (
                // eslint-disable-next-line jsx-a11y/click-events-have-key-events,jsx-a11y/no-static-element-interactions
                <div
                    className={classNames(`${CLASSNAME}__after`, {
                        [`${CLASSNAME}__after--is-clickable`]: hasAfterClick,
                    })}
                    onClick={handleOnAfterClick}
                >
                    {after}
                </div>
            )}
        </a>
    );
});
Chip.displayName = COMPONENT_NAME;
Chip.className = CLASSNAME;
Chip.defaultProps = DEFAULT_PROPS;
