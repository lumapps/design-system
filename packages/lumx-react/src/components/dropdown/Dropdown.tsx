import React, { cloneElement, forwardRef, useMemo, useRef } from 'react';

import classNames from 'classnames';

import { List, ListProps } from '@lumx/react/components/list/List';
import { Offset, Placement, Popover } from '@lumx/react/components/popover/Popover';
import { useInfiniteScroll } from '@lumx/react/hooks/useInfiniteScroll';
import { Comp, GenericProps, getRootClassName, handleBasicClasses, isComponent } from '@lumx/react/utils';

/**
 * Defines the props of the component.
 */
export interface DropdownProps extends GenericProps {
    /** Reference to the element around which the dropdown is placed.
     * @see {@link PopoverProps#anchorRef}
     */
    anchorRef: React.RefObject<HTMLElement>;
    /** Dropdown content. */
    children: React.ReactNode;
    /**
     * Whether a click anywhere out of the Dropdown would close it or not.
     * @see {@link PopoverProps#closeOnClickAway}
     */
    closeOnClickAway?: boolean;
    /**
     * Whether to close the Dropdown when clicking in it or not.
     */
    closeOnClick?: boolean;
    /**
     * Whether an escape key press would close the Dropdown or not.
     * @see {@link PopoverProps#closeOnEscape}
     */
    closeOnEscape?: boolean;
    /**
     * Whether the dropdown should fit to the anchor width (if dropdown is smaller) or not.
     * @see {@link PopoverProps#fitToAnchorWidth}
     */
    fitToAnchorWidth?: boolean;
    /**
     * Whether the dropdown should shrink to fit within the viewport height or not.
     * @see {@link PopoverProps#fitWithinViewportHeight}
     */
    fitWithinViewportHeight?: boolean;
    /**
     * Whether the dropdown should be displayed or not. Useful to control the Dropdown from outside the component.
     * @see {@link PopoverProps#isOpen}
     */
    isOpen: boolean;
    /**
     * Offset applied to the Dropdown position.
     * @see {@link PopoverProps#offset}
     */
    offset?: Offset;
    /**
     * Preferred Dropdown placement against the anchor element.
     * @see {@link PopoverProps#placement}
     */
    placement?: Placement;
    /** Whether the focus should be set on the list when the dropdown is open or not. */
    shouldFocusOnOpen?: boolean;
    /**
     * Z-axis position.
     * @see {@link PopoverProps#zIndex}
     */
    zIndex?: number;
    /**
     * On close callback.
     * @see {@link PopoverProps#onClose}
     */
    onClose?(): void;
    /** On scroll end callback. */
    onInfiniteScroll?(): void;
}

/**
 * Component display name.
 */
const COMPONENT_NAME = 'Dropdown';

/**
 * Component default class name and class prefix.
 */
const CLASSNAME = getRootClassName(COMPONENT_NAME);

/**
 * Component default props.
 */
const DEFAULT_PROPS: Partial<DropdownProps> = {
    closeOnClick: true,
    closeOnClickAway: true,
    closeOnEscape: true,
    fitToAnchorWidth: true,
    fitWithinViewportHeight: true,
    placement: Placement.BOTTOM_START,
    shouldFocusOnOpen: true,
};

/**
 * Dropdown component.
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */
export const Dropdown: Comp<DropdownProps, HTMLDivElement> = forwardRef((props, ref) => {
    const {
        anchorRef,
        children,
        className,
        closeOnClick,
        closeOnClickAway,
        closeOnEscape,
        fitToAnchorWidth,
        fitWithinViewportHeight,
        isOpen,
        offset,
        onClose,
        onInfiniteScroll,
        placement,
        shouldFocusOnOpen,
        zIndex,
        ...forwardedProps
    } = props;
    const innerRef = useRef<HTMLDivElement>(null);
    const listElement = useRef(null);

    useInfiniteScroll(innerRef, onInfiniteScroll);

    const popperElement = useMemo(() => {
        return !Array.isArray(children) && isComponent(List)(children)
            ? cloneElement<ListProps>(children, {
                  ...children.props,
                  ref: listElement,
                  onClick(evt: MouseEvent) {
                      children.props.onClick?.(evt);

                      if (closeOnClick) {
                          onClose?.();
                      }
                  },
                  isClickable: true,
              })
            : children;
    }, [children, closeOnClick, onClose]);

    return isOpen ? (
        <Popover
            ref={ref}
            {...forwardedProps}
            anchorRef={anchorRef}
            className={classNames(className, handleBasicClasses({ prefix: CLASSNAME }))}
            elevation={0 as any}
            closeOnClickAway={closeOnClickAway}
            closeOnEscape={closeOnEscape}
            fitToAnchorWidth={fitToAnchorWidth}
            fitWithinViewportHeight={fitWithinViewportHeight}
            focusElement={shouldFocusOnOpen ? listElement : undefined}
            isOpen={isOpen}
            offset={offset}
            onClose={onClose}
            placement={placement}
            zIndex={zIndex}
        >
            <div className={`${CLASSNAME}__menu`} ref={innerRef}>
                {popperElement}
            </div>
        </Popover>
    ) : null;
});
Dropdown.displayName = COMPONENT_NAME;
Dropdown.className = CLASSNAME;
Dropdown.defaultProps = DEFAULT_PROPS;
