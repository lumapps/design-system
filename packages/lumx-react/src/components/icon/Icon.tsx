import React, { forwardRef } from 'react';

import classNames from 'classnames';

import { Color, ColorPalette, ColorVariant, Size, Theme } from '@lumx/react';
import { Comp, GenericProps, getRootClassName, handleBasicClasses } from '@lumx/react/utils';

export type IconSizes = Extract<Size, 'xxs' | 'xs' | 's' | 'm' | 'l' | 'xl' | 'xxl'>;

/**
 * Defines the props of the component.
 */
export interface IconProps extends GenericProps {
    /** Color variant. */
    color?: Color;
    /** Lightened or darkened variant of the selected icon color. */
    colorVariant?: ColorVariant;
    /** Whether the icon has a shape. */
    hasShape?: boolean;
    /**
     * Icon (SVG path).draw code (`d` property of the `<path>` SVG element).
     * @see {@link https://developer.mozilla.org/en-US/docs/Web/SVG/Tutorial/Paths}
     */
    icon: string;
    /** Size variant. */
    size?: IconSizes;
    /** Theme adapting the component to light or dark background. */
    theme?: Theme;
}

/**
 * Component display name.
 */
const COMPONENT_NAME = 'Icon';

/**
 * Component default class name and class prefix.
 */
const CLASSNAME = getRootClassName(COMPONENT_NAME);

/**
 * Component default props.
 */
const DEFAULT_PROPS: Partial<IconProps> = {};

/**
 * Icon component.
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */
export const Icon: Comp<IconProps, HTMLElement> = forwardRef((props, ref) => {
    const { className, color, colorVariant, hasShape, icon, size, theme, ...forwardedProps } = props;
    let iconColor;
    if (color) {
        iconColor = color;
    } else if (theme) {
        iconColor = theme === Theme.light ? ColorPalette.dark : ColorPalette.light;
    } else if (hasShape) {
        iconColor = ColorPalette.dark;
    }

    let iconSize;
    if (size) {
        if (hasShape) {
            if (size === Size.xxs || size === Size.xs) {
                iconSize = Size.s;
            } else if (size === Size.xxl) {
                iconSize = Size.xl;
            } else {
                iconSize = size;
            }
        } else {
            iconSize = size;
        }
    } else if (hasShape) {
        iconSize = Size.m;
    }

    return (
        <i
            ref={ref}
            {...forwardedProps}
            className={classNames(
                className,
                handleBasicClasses({
                    color: iconColor,
                    colorVariant,
                    hasShape,
                    prefix: CLASSNAME,
                    size: iconSize,
                }),
                !hasShape && `${CLASSNAME}--no-shape`,
                `${CLASSNAME}--path`,
            )}
        >
            <svg
                aria-hidden="true"
                height="1em"
                preserveAspectRatio="xMidYMid meet"
                style={{ verticalAlign: '-0.125em' }}
                viewBox="0 0 24 24"
                width="1em"
            >
                <path d={icon} fill="currentColor" />
            </svg>
        </i>
    );
});
Icon.displayName = COMPONENT_NAME;
Icon.className = CLASSNAME;
Icon.defaultProps = DEFAULT_PROPS;
