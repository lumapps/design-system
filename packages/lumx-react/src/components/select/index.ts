export * from './Select';
export * from './SelectMultiple';
