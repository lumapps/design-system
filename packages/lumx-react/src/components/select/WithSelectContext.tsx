import React, { Ref, useCallback, useMemo, useRef } from 'react';

import classNames from 'classnames';
import { uid } from 'uid';

import { Kind, Theme } from '@lumx/react/components';
import { Dropdown } from '@lumx/react/components/dropdown/Dropdown';
import { InputHelper } from '@lumx/react/components/input-helper/InputHelper';
import { Placement } from '@lumx/react/components/popover/Popover';

import { getRootClassName, handleBasicClasses } from '@lumx/react/utils';
import { mergeRefs } from '@lumx/react/utils/mergeRefs';
import { useListenFocus } from '@lumx/react/hooks/useListenFocus';
import { CoreSelectProps, SelectVariant } from './constants';

/** The display name of the component. */
const COMPONENT_NAME = 'Select';

/** The default class name and classes prefix for this component. */
const CLASSNAME = getRootClassName(COMPONENT_NAME);

/** The default value of props. */
export const DEFAULT_PROPS: Partial<CoreSelectProps> = {
    theme: Theme.light,
    variant: SelectVariant.input,
};

export const WithSelectContext = (
    SelectElement: React.FC<any>,
    {
        children,
        className,
        isMultiple,
        closeOnClick = !isMultiple,
        disabled,
        error,
        hasError,
        helper,
        id,
        isDisabled = disabled,
        isEmpty,
        isOpen,
        isRequired,
        isValid,
        label,
        onClear,
        onDropdownClose,
        onInfiniteScroll,
        onInputClick,
        placeholder,
        theme = DEFAULT_PROPS.theme,
        value,
        variant = DEFAULT_PROPS.variant,
        ...forwardedProps
    }: CoreSelectProps,
    ref: Ref<HTMLDivElement>,
): React.ReactElement => {
    const selectId = useMemo(() => id || `select-${uid()}`, [id]);
    const anchorRef = useRef<HTMLElement>(null);
    const selectRef = useRef<HTMLDivElement>(null);
    const isFocus = useListenFocus(anchorRef);

    const handleKeyboardNav = useCallback(
        (evt: React.KeyboardEvent<HTMLElement>) => {
            if ((evt.key === 'Enter' || evt.key === ' ' || evt.key === 'ArrowDown') && onInputClick) {
                evt.preventDefault();
                onInputClick();
            }
        },
        [onInputClick],
    );

    const onClose = () => {
        if (onDropdownClose) {
            onDropdownClose();
        }
        anchorRef?.current?.blur();
    };

    return (
        <div
            ref={mergeRefs(ref, selectRef)}
            className={classNames(
                className,
                handleBasicClasses({
                    hasError,
                    hasLabel: Boolean(label),
                    hasPlaceholder: Boolean(placeholder),
                    hasValue: !isEmpty,
                    isDisabled,
                    isEmpty,
                    isFocus,
                    isOpen,
                    isValid,
                    prefix: CLASSNAME,
                    theme: theme === Theme.light ? Theme.light : Theme.dark,
                }),
            )}
        >
            <SelectElement
                {...forwardedProps}
                anchorRef={anchorRef}
                aria-disabled={isDisabled}
                handleKeyboardNav={handleKeyboardNav}
                hasError={hasError}
                isDisabled={isDisabled}
                isEmpty={isEmpty}
                isRequired={isRequired}
                isValid={isValid}
                label={label}
                placeholder={placeholder}
                id={selectId}
                theme={theme}
                value={value}
                variant={variant}
                onClear={onClear}
                onInputClick={onInputClick}
            />
            <Dropdown
                anchorRef={anchorRef}
                closeOnClick={closeOnClick}
                closeOnClickAway
                closeOnEscape
                isOpen={!!isOpen}
                placement={Placement.BOTTOM_START}
                onClose={onClose}
                onInfiniteScroll={onInfiniteScroll}
            >
                {children}
            </Dropdown>
            {hasError && error && (
                <InputHelper className={`${CLASSNAME}__helper`} kind={Kind.error} theme={theme}>
                    {error}
                </InputHelper>
            )}
            {helper && (
                <InputHelper className={`${CLASSNAME}__helper`} theme={theme}>
                    {helper}
                </InputHelper>
            )}
        </div>
    );
};
