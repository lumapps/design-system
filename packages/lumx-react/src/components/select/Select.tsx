import React, { forwardRef, RefObject } from 'react';

import classNames from 'classnames';
import lodashIsEmpty from 'lodash/isEmpty';

import { mdiAlertCircle, mdiCheckCircle, mdiCloseCircle, mdiMenuDown } from '@lumx/icons';

import { Emphasis, Size } from '@lumx/react/components';
import { IconButton } from '@lumx/react/components/button/IconButton';
import { Chip } from '@lumx/react/components/chip/Chip';
import { Icon } from '@lumx/react/components/icon/Icon';
import { InputLabel } from '@lumx/react/components/input-label/InputLabel';

import { Comp, getRootClassName, handleBasicClasses } from '@lumx/react/utils';

import { WithSelectContext } from './WithSelectContext';
import { CoreSelectProps, SelectVariant } from './constants';

/** Defines the props of the component. */
export interface SelectProps extends CoreSelectProps {
    /** Selected value. */
    value: string;
}

export { SelectVariant };

/** The display name of the component. */
const COMPONENT_NAME = 'Select';

/** The default class name and classes prefix for this component. */
const CLASSNAME = getRootClassName(COMPONENT_NAME);

/** The default value of props. */
const DEFAULT_PROPS: Partial<SelectProps> = {
    selectedValueRender: (choice) => choice,
};

const stopPropagation = (evt: Event) => evt.stopPropagation();

/**
 * Select component.
 */
const SelectField: React.FC<SelectProps> = ({
    anchorRef,
    clearButtonProps,
    handleKeyboardNav,
    hasError,
    hasInputClear,
    id,
    isDisabled,
    isEmpty,
    isRequired,
    isValid,
    label,
    onClear,
    onInputClick,
    placeholder,
    selectedValueRender,
    theme,
    value,
    variant,
    ...forwardedProps
}) => {
    return (
        <>
            {variant === SelectVariant.input && (
                <>
                    {label && (
                        <div className={`${CLASSNAME}__header`}>
                            <InputLabel
                                htmlFor={id}
                                className={`${CLASSNAME}__label`}
                                isRequired={isRequired}
                                theme={theme}
                            >
                                {label}
                            </InputLabel>
                        </div>
                    )}

                    {/* eslint-disable-next-line jsx-a11y/no-static-element-interactions */}
                    <div
                        ref={anchorRef as RefObject<HTMLDivElement>}
                        id={id}
                        className={`${CLASSNAME}__wrapper`}
                        onClick={onInputClick}
                        onKeyDown={handleKeyboardNav}
                        tabIndex={isDisabled ? undefined : 0}
                        aria-disabled={isDisabled || undefined}
                        {...forwardedProps}
                    >
                        <div
                            className={classNames([
                                `${CLASSNAME}__input-native`,
                                isEmpty && placeholder && `${CLASSNAME}__input-native--placeholder`,
                            ])}
                        >
                            {!isEmpty && <span>{selectedValueRender?.(value)}</span>}

                            {isEmpty && placeholder && <span>{placeholder}</span>}
                        </div>

                        {(isValid || hasError) && (
                            <div className={`${CLASSNAME}__input-validity`}>
                                <Icon icon={isValid ? mdiCheckCircle : mdiAlertCircle} size={Size.xxs} />
                            </div>
                        )}

                        {hasInputClear && clearButtonProps && (
                            <IconButton
                                {...clearButtonProps}
                                className={`${CLASSNAME}__input-clear`}
                                icon={mdiCloseCircle}
                                emphasis={Emphasis.low}
                                size={Size.s}
                                theme={theme}
                                onClick={onClear}
                                onKeyDown={stopPropagation}
                            />
                        )}

                        <div className={`${CLASSNAME}__input-indicator`}>
                            <Icon icon={mdiMenuDown} size={Size.s} />
                        </div>
                    </div>
                </>
            )}

            {variant === SelectVariant.chip && (
                <Chip
                    id={id}
                    isSelected={!isEmpty}
                    isDisabled={isDisabled}
                    after={<Icon icon={isEmpty ? mdiMenuDown : mdiCloseCircle} />}
                    onAfterClick={isEmpty ? onInputClick : onClear}
                    onClick={onInputClick}
                    ref={anchorRef as RefObject<HTMLAnchorElement>}
                    theme={theme}
                    {...forwardedProps}
                >
                    {isEmpty && <span>{label}</span>}

                    {!isEmpty && <span>{selectedValueRender?.(value)}</span>}
                </Chip>
            )}
        </>
    );
};

/**
 * Select component.
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */
export const Select: Comp<SelectProps, HTMLDivElement> = forwardRef((props, ref) => {
    const isEmpty = lodashIsEmpty(props.value);
    const hasInputClear = props.onClear && props.clearButtonProps && !isEmpty;

    return WithSelectContext(
        SelectField,
        {
            ...props,
            className: classNames(
                props.className,
                handleBasicClasses({
                    hasInputClear,
                    hasUnique: !props.isEmpty,
                    prefix: CLASSNAME,
                }),
            ),
            hasInputClear,
            isEmpty,
        },
        ref,
    );
});
Select.displayName = COMPONENT_NAME;
Select.className = CLASSNAME;
Select.defaultProps = DEFAULT_PROPS;
Select.className = CLASSNAME;
