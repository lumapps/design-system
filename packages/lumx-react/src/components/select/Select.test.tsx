import React, { ReactElement, RefObject } from 'react';

import { mount, ReactWrapper, shallow, ShallowWrapper } from 'enzyme';
import 'jest-enzyme';

import { Kind, Theme } from '@lumx/react/components';
import { Chip } from '@lumx/react/components/chip/Chip';
import { Icon } from '@lumx/react/components/icon/Icon';

import { commonTestsSuite, Wrapper } from '@lumx/react/testing/utils';
import { getBasicClass } from '@lumx/react/utils';

import { mdiCloseCircle, mdiMenuDown } from '@lumx/icons';
import { Dropdown } from '@lumx/react/components/dropdown/Dropdown';
import { Select, SelectProps, SelectVariant } from './Select';
import { DEFAULT_PROPS } from './WithSelectContext';

const CLASSNAME = Select.className as string;

jest.mock('uid', () => ({ uid: () => 'uid' }));

type SetupProps = Partial<SelectProps>;

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 */
const setup = ({ ...propsOverride }: SetupProps = {}, shallowRendering = true) => {
    const props: SelectProps = {
        children: <span>Select Component</span>,
        value: '',
        ...propsOverride,
    };

    const renderer: (el: ReactElement) => Wrapper = shallowRendering ? shallow : mount;
    const wrapper: ShallowWrapper | ReactWrapper = renderer(<Select {...props} />);

    return {
        container: wrapper.find('div').first(),
        dropdown: wrapper.find(Dropdown),
        error: wrapper
            .findWhere(
                (n: ShallowWrapper | ReactWrapper) => n.name() === 'InputHelper' && n.prop('kind') === Kind.error,
            )
            .first(),
        helper: wrapper.findWhere(
            (n: ShallowWrapper | ReactWrapper) => n.name() === 'InputHelper' && n.prop('kind') === Kind.info,
        ),
        input: wrapper.find('#select-uid:not(SelectField)').first(),
        chip: wrapper.find('Chip'),
        inputWrapper: wrapper.find('.lumx-select__wrapper'),
        props,
        wrapper,
    };
};

describe(`<${Select.displayName}>`, () => {
    // 1. Test render via snapshot (default states of component).
    describe('Snapshots and structure', () => {
        it('should render correctly', () => {
            const { container, wrapper } = setup();
            expect(wrapper).toMatchSnapshot();

            expect(container).toExist();
            expect(container).toHaveClassName(CLASSNAME);
        });
    });

    // 2. Test defaultProps value and important props custom values.
    describe('Props', () => {
        it('should have default classNames', () => {
            const { wrapper, container } = setup();
            wrapper.update();

            expect(container).toHaveClassName(CLASSNAME);
            expect(container).toHaveClassName(
                getBasicClass({ prefix: CLASSNAME, type: 'theme', value: DEFAULT_PROPS.theme }),
            );
            expect(container).toHaveClassName(getBasicClass({ prefix: CLASSNAME, type: 'isEmpty', value: true }));
        });

        it('should use the given `theme`', () => {
            const testedProp = 'theme';
            const modifiedProps: SetupProps = {
                [testedProp]: Theme.dark,
            };

            const { container } = setup(modifiedProps);

            expect(container).toHaveClassName(
                getBasicClass({ prefix: CLASSNAME, type: testedProp, value: modifiedProps[testedProp] }),
            );
            expect(container).not.toHaveClassName(
                getBasicClass({ prefix: CLASSNAME, type: testedProp, value: DEFAULT_PROPS.theme }),
            );
        });

        it('should use the given `isValid`', () => {
            const testedProp = 'isValid';
            const modifiedProps: SetupProps = {
                [testedProp]: true,
            };

            const { container } = setup(modifiedProps);

            expect(container).toHaveClassName(
                getBasicClass({ prefix: CLASSNAME, type: testedProp, value: modifiedProps[testedProp] }),
            );
        });

        it('should use the given `hasError`', () => {
            const testedProp = 'hasError';
            const modifiedProps: SetupProps = {
                [testedProp]: true,
            };

            const { container } = setup(modifiedProps);

            expect(container).toHaveClassName(
                getBasicClass({ prefix: CLASSNAME, type: testedProp, value: modifiedProps[testedProp] }),
            );
        });

        it('should display the given `error`', () => {
            const { container, error } = setup({
                error: 'You are not bold!',
                hasError: true,
            });

            expect(error).toExist();
            expect(container).toHaveClassName(getBasicClass({ prefix: CLASSNAME, type: 'hasError', value: true }));
        });

        it('should NOT display the given `error`', () => {
            const { container, error } = setup({
                error: 'You are not bold!',
                hasError: false,
            });

            expect(error).not.toExist();
            expect(container).not.toHaveClassName(getBasicClass({ prefix: CLASSNAME, type: 'hasError', value: true }));
        });

        it('should use the given `value`', () => {
            const testedProp = 'value';
            const modifiedProps: SetupProps = {
                [testedProp]: 'val1',
            };

            const { container } = setup(modifiedProps);

            expect(container).toHaveClassName(getBasicClass({ prefix: CLASSNAME, type: 'hasUnique', value: true }));
            expect(container).toHaveClassName(getBasicClass({ prefix: CLASSNAME, type: 'hasValue', value: true }));
            expect(container).not.toHaveClassName(getBasicClass({ prefix: CLASSNAME, type: 'isEmpty', value: true }));
        });

        it('should trigger the given `onDropdownClose` on dropdown close', () => {
            const onDropdownClose = jest.fn();
            const { dropdown } = setup({ onDropdownClose });

            const props: any = dropdown.props();

            props.onClose();

            expect(onDropdownClose).toHaveBeenCalled();
        });

        it('should pass the given `isOpen` to the dropdown', () => {
            const isOpen = true;
            const { dropdown } = setup({ isOpen });

            expect(dropdown).toHaveProp('isOpen', isOpen);
        });

        it('should display the given `helper`', () => {
            const { helper } = setup({
                helper: 'Be bold',
            });

            expect(helper).toExist();
        });

        it('should display the given `helper` and `error`', () => {
            const { error, helper } = setup({
                error: 'You are not bold!',
                hasError: true,
                helper: 'Be bold',
            });

            expect(error).toExist();
            expect(helper).toExist();
        });

        it('should have a data-id as prop', () => {
            const { inputWrapper } = setup(
                {
                    'data-id': 'select',
                },
                false,
            );

            expect(inputWrapper.prop('data-id')).toEqual('select');
        });

        it('should have a data-id as prop with Chip variant', () => {
            const { chip } = setup(
                {
                    'data-id': 'select',
                    variant: SelectVariant.chip,
                },
                false,
            );

            expect(chip.prop('data-id')).toEqual('select');
        });
    });

    // 3. Test events.
    describe('Events', () => {
        const onClick: jest.Mock = jest.fn();

        beforeEach(() => {
            onClick.mockClear();
        });

        describe('should trigger `onInputClick` when the select button is clicked', () => {
            it('with input variant', () => {
                const { input } = setup({ onInputClick: onClick, variant: SelectVariant.input }, false);

                input.simulate('click');

                expect(onClick).toHaveBeenCalled();
            });

            it('with chip variant', () => {
                const { input } = setup({ onInputClick: onClick, variant: SelectVariant.chip }, false);

                input.simulate('click');

                expect(onClick).toHaveBeenCalled();
            });
        });

        it('should call onClear when clear icon is clicked in select input', () => {
            const value = 'Value';
            const onClear: jest.Mock = jest.fn();
            const { input } = setup({ value, onClear, clearButtonProps: { label: 'Clear' } }, false);

            input.find('[aria-label="Clear"]').first().simulate('click');
            expect(onClear).toHaveBeenCalled();
        });
    });

    // 4. Test conditions (i.e. things that display or not in the UI based on props).
    describe('Conditions', () => {
        describe('Input variant', () => {
            it('should render a div as the select input', () => {
                const { dropdown, input } = setup({}, false);
                const dropdownRef: RefObject<HTMLDivElement> = dropdown.prop('anchorRef');
                expect(input.type()).toBe('div');
                // Impossible to get the ref passed to a div so we check if ref.current is matching the input html.
                expect(input.html()).toEqual(dropdownRef && dropdownRef.current && dropdownRef.current.outerHTML);
            });

            describe('Has value', () => {
                const value = 'Value';
                const hasValueProps: Partial<SetupProps> = {
                    value,
                    variant: SelectVariant.input,
                };

                it('should render the value selected if not multiple ', () => {
                    const { input } = setup({ ...hasValueProps }, false);

                    expect(input.find(`.${CLASSNAME}__input-native span`).first().text()).toBe(value);
                });
            });

            describe('No value', () => {
                const placeholder = 'My placeholder';
                const hasNoValueProps: Partial<SetupProps> = {
                    placeholder,
                    value: '',
                    variant: SelectVariant.input,
                };

                it('should render the placeholder when empty', () => {
                    const { input } = setup({ ...hasNoValueProps }, false);

                    expect(input.find(`.${CLASSNAME}__input-native--placeholder span`).first().text()).toBe(
                        placeholder,
                    );
                });
            });
        });

        describe('Chip variant', () => {
            it('should render a Chip as the select input', () => {
                const { input } = setup({ variant: SelectVariant.chip }, false);

                expect(input.type()).toEqual(Chip);
            });

            describe('Has value', () => {
                const value = 'Value';
                const hasValueProps: Partial<SetupProps> = {
                    value,
                    variant: SelectVariant.chip,
                };

                it('should render the value', () => {
                    const label = 'The label';
                    const { input } = setup(
                        {
                            ...hasValueProps,
                            label,
                        },
                        false,
                    );

                    expect(input.find('.lumx-chip__label').text()).toEqual(value);
                });

                it('should render a close icon if there is a value', () => {
                    const { input } = setup({ ...hasValueProps }, false);

                    expect(input.find(Icon).first().prop('icon')).toBe(mdiCloseCircle);
                });
            });

            describe('No value', () => {
                const hasNoValue: Partial<SetupProps> = {
                    value: '',
                    variant: SelectVariant.chip,
                };

                it('should render the label', () => {
                    const label = 'The label';
                    const { input } = setup({ ...hasNoValue, label }, false);

                    expect(input.find('.lumx-chip__label').text()).toEqual(label);
                });

                it('should render a MenuDown icon', () => {
                    const { input } = setup({ ...hasNoValue }, false);

                    expect(input.find(Icon).first().prop('icon')).toBe(mdiMenuDown);
                });
            });
        });
    });

    // 5. Test state.
    describe('State', () => {
        // Nothing to do here.
    });

    // Common tests suite.
    commonTestsSuite(setup, { className: 'wrapper' }, { className: CLASSNAME });
});
