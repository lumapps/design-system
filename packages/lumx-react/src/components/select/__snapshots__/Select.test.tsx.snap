// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select> Snapshots and structure should render correctly 1`] = `
<div
  className="lumx-select lumx-select--has-unique lumx-select lumx-select--is-empty lumx-select--theme-light"
>
  <SelectField
    anchorRef={
      Object {
        "current": null,
      }
    }
    handleKeyboardNav={[Function]}
    id="select-uid"
    isEmpty={true}
    selectedValueRender={[Function]}
    theme="light"
    value=""
    variant="input"
  />
  <Dropdown
    anchorRef={
      Object {
        "current": null,
      }
    }
    closeOnClick={true}
    closeOnClickAway={true}
    closeOnEscape={true}
    fitToAnchorWidth={true}
    fitWithinViewportHeight={true}
    isOpen={false}
    onClose={[Function]}
    placement="bottom-start"
    shouldFocusOnOpen={true}
  >
    <span>
      Select Component
    </span>
  </Dropdown>
</div>
`;
